source/mainnet/smart-contracts/best-practices/development.rst

```diff
@@ -0,0 +1,203 @@
+
+- Determine the “bare minimum” that should be on-chain: use smart contracts only for the part that requires decentralization.
+- Keep your contracts simple and minimalistic: stick to strictly necessary functionality only.
+  The more complexity you have in your contract, the larger the attack surface.
+- Be ready if things go wrong after deployment:
+
+  - provide “pause” functionality;
+  - implement additional approvals/wait periods for dangerous operations;
+  - have a clear plan for how to fix bugs found after deployment.
+
+- Extensively review, test and apply automated analysis/verification tools. Use different methods to ensure the correctness of your code.
+  Do not trust one particular method or tool; make sure that many people have looked at the code and interacted with the smart contract.
+
+
+.. _best-practices-specification:
+
+Specification
+=============
+
+A smart contract specification reflects developers' intentions regarding the smart contract functionality.
+It can serve as a guide for the implementation as well as for testing, verifying, and auditing the code.
+Start with an outline of the functionality and gradually refine it into specifications of contract entrypoints.
+
+For the entrypoints, consider the following:
+
+- What is the interface (entrypoints and their parameters)? For example, ``transfer`` takes three parameters ``from``, ``to`` addresses, and ``amount``.
+- Who can access the entrypoints? For example, the transaction sender must be an operator of the ``from`` account.
+- Which properties of these entrypoints should be satisfied? For example, a successful call of ``transfer(from, to, amount)`` decreases the ``from`` balance and the ``to`` balance by ``amount``.
+
+If the logic of your application requires complex flows that include several contract calls, or features several interacting smart contracts, use diagrams with descriptions to document these.
+
+The specification is not static; it can be refined once you start developing your application.
+Work in iterations, refining the specification when new requirements are discovered or old requirements change.
+
+An informal specification can be refined, turned into a formal one, and used as input to automated testing and verification tools.
+
+The absence of specifications complicates the quality assurance of smart contracts.
+For example, it makes the auditing process less efficient and more expensive, because the specification has to be recovered from the source code before starting the audit process.
+
+
+Concordium Rust Smart Contracts
+===============================
+
+This section provides recommendations for developing smart contracts in Rust.
+
+
+.. _best-practices-code-structure:
+
+Recommended structure
+---------------------
+
+- Use ``cargo init`` with an appropriate template to start a new project.
+- Build contract logic as the state struct implementation.
```

Comment:
Is this what we recommend?

Reply:
I've talked to @abizjak about this at some point, and my understanding was that this is a recommended structure. Most of our non-trivial examples follow this approach. But I might be wrong about this :) So, I'm open to other suggestions.

Reply:
I mean it is fine and we do use it. I was just unaware that it was now the recommended approach ;)